sway-lib-std/src/storage.sw

```diff
@@ -40,21 +39,18 @@ use ::result::Result;
             local_key = sha256(local_key);
         }
 
-        // Store the leftover bytes using a single `swwq`
-        asm(k: local_key, v: ptr_to_value) {
-            swwq k v;
-        };
+        // Store the leftover bytes using a single quad store
+        __state_store_quad(local_key, ptr_to_value);
     };
 }
 
 /// Load a stack variable from storage.
 #[storage(read)]pub fn get<T>(key: b256) -> T {
     if !__is_reference_type::<T>() {
-        // If copy type, then it's a single word and can be read with a single
-        // SRW.
-        asm(k: key, v) {
-            srw v k;
-            v: T
+        // If copy type, then it's a single word
+        let loaded_word = __state_load_word(key);
+        asm (l: loaded_word) {
+            l: T
         }
```

Comment:
Can we not turbofish this one here?  Was that also part of the discussion around this issue?

Reply:
> Can we not turbofish this one here? Was that also part of the discussion around this issue?

You mean using `std::mem::read<T>`? That wouldn't work as it ends up actually doing a load, assuming the value to be an address (in case of non-reference types).

Reply:
The asm block is just casting the loaded word to `T` -- if it's a copy type then we should be able to use `__state_load_word::<T>(key)` to get it in the right type from the start, no?  I didn't check, is `__state_load_word()` just returning a `u64` right now?  Maybe it should return `T`.

Reply:
Or is this just the problem we can't solve until we have trait constraints?

Reply:
It really only returns `u64`, so having its type to return `T` would be incorrect. But if it's returning only `u64`, then when used with `is_reference_type<T>` in a generic function, we end up with this problem we have. If we could restrict it with trait contraints, then we wouldn't have this call in a non `u64` context at all.

Reply:
Right, OK.  I kinda feel like we should only have two intrinsics -- read and write, and whether they need to use `SRW`/`SWW` or `SRWQ`/`SWWQ` can be decided by the back end.  Then the read functions _could_ just return `T`.

But I'm sure that won't work either for some reason.  OK, this'll do for now but I can feel the technical debt growing weekly. :smile: